### DISCUSSION TITLE
Server-side islands with `server:dynamic`

### DISCUSSION THREAD

**Discussion by Tc-001** [withastro/roadmap#116] [Proposal]
With the ssr experimental flag released, I propose an idea I have had bouncing around in my head.

# Propsal

It may not always be necessary to render all components, especially if you have to await API network requests or use heavy libraries server-side. 
Doing so on each request would waste billable server time with serverless, slow down the page loads for the user and SEO crawlers, as well as exceed API quotas if the developer is not prepared for a sudden influx.

My idea is a `server:` directive that behaves similarly to the `client:` one already in use. 

By default, all components are rendered statically on server build, but they can be toggled to be SSR by using `server:dynamic`.

# Example

```astro
---
import CurrentTime from "./foo.astro"
import SlowComponent from "./bar.astro"
import SlowToBuildSvelte from "./baz.svelte"
import SmallComponent from "./bax.svalte"
---

<body>
    <CurrentTime server:dynamic /> <!-- Will render on request -->
    <SlowComponent /> <!-- Statically redered -->
    <SlowToBuildSvelte client:load /> <!-- Statically redered but still reactive -->
	<SmallComponent client:load server:dynamic /> <!-- (May not be possible) Rendered on request and reactive -->
</body>
```

# Syntax

In my proposal `server:` will have only one mode, that being `server:dynamic`. 

# Open questions

Would it be useful to have other server-side modes too, for example `server:cache={time ms}`?

**Comment by matthewp** [comment]
This is very interesting idea! I particularly like how it aligns with the island architecture that Astro users are already familiar with.

Question: What about the page component? I guess it can't be an island by itself, and you need something in case of redirects or 404s. So I'm curious if in this model the page component would have some other way to signal that it needs to 404.

**Reply by Tc-001** [reply]
To signal that a certain page is the 404 component, you could just have the file named `404.astro` as it is done now.

(If I understood the question correctly) If you want the whole page to be SSR I see some options:

* Empty page with the actual content as a component
* `<Fragment server:dynamic>` (don't know how possible)
* Have it as a config option (ex. `"noStaticSSR" : true` for the whole site to be dynamic)
* If using adapters, setting specific paths in their config could be an option

**Reply by KaninchenSpeed** [reply]
Maybe an option to cache the rendered page for the specific combinations of url query arguments. Or even a condition callback which returns true if the cached page should be diaplayed or false if the page should be rerendered.

**Comment by kurtextrem** [comment]
Cross-linking this Twitter thread so it's not lost on X: https://twitter.com/matthewcp/status/1714677585857114283

**Comment by matthewp** [comment]
This proposal has been accepted as is now stage 2: https://github.com/withastro/roadmap/issues/945

Please continue the discussion there.
